### PR TITLE
Add unsubscribe link to event emails

### PR DESCRIPTION
**Pull request by kdeloach**
Connects #115
Connects #1660

### CODE REVIEW

**Comment by kdeloach** [comment]
This is waiting for feedback from #115.

**Comment by kdeloach** [comment]
@jwalgran @hectcastro This has been updated to display a friendly error message if someone tries to use an invalid or expired token. Also, event emails are always sent from the default configured sender instead of the group contact.

File: src/nyc_trees/apps/event/templates/event/email_rsvps.html [hunk 1]
```diff
@@ -33,12 +33,12 @@ <h4 class="modal-title">Event Registrants</h4>
             </div>
```
Comment (hectcastro) [inline src/nyc_trees/apps/event/templates/event/email_rsvps.html]: Potentially unrelated to this PR, but I'm getting an `XMLHttpRequest` cross-origin request error on the "See all" link of this page.

File: src/nyc_trees/apps/core/models.py [hunk 2]
```diff
@@ -227,6 +229,18 @@ def blocks_mapped_count(self):
...
         return url_if_cooked(self.reservations_map_pdf_filename)
 
+    def make_token(self):
+        return TimestampSigner().sign(self.username)
+
+    # Source: http://grokcode.com/819/one-click-unsubscribes-for-django-apps/
+    def is_valid_token(self, token):
+        try:
+            # Valid for 2 days
+            TimestampSigner().unsign(token, max_age=60 * 60 * 48)
```
Comment (hectcastro) [inline src/nyc_trees/apps/core/models.py]: After thinking about this over lunch, I couldn't think of a good reason to bound the token with a TTL. Not something that needs to be addressed here though.

**Comment by kdeloach** [comment]
I've added another commit to this PR for issue #1660.

File: src/nyc_trees/apps/mail/libs.py [hunk 3]
```diff
@@ -36,10 +36,22 @@ def send_to(user, message_type, *args, **kwargs):
...
     to = user.email
 
-    msg = EmailMessage(subject, body_text, from_email, [to])
+    headers = {}
+    if reply_to:
+        headers.update({
+            'Reply-To': reply_to
+        })
+
+    msg = EmailMessage(subject, body_text, from_email, [to], headers=headers)
```
Comment (hectcastro) [inline src/nyc_trees/apps/mail/libs.py]: This is missing a list for BCC addresses after `[to]`.

File: src/nyc_trees/apps/event/templates/event/email_rsvps.html (hunk 1 above)
Comment (kdeloach) [inline src/nyc_trees/apps/event/templates/event/email_rsvps.html]: Interesting, I thought it just opens a static Bootstrap modal.

File: src/nyc_trees/apps/core/models.py (hunk 2 above)
Comment (kdeloach) [inline src/nyc_trees/apps/core/models.py]: True, this may be overkill.

File: src/nyc_trees/apps/mail/libs.py (hunk 3 above)
Comment (kdeloach) [inline src/nyc_trees/apps/mail/libs.py]: It's optional according to the API docs.

File: src/nyc_trees/apps/event/templates/event/email_rsvps.html (hunk 1 above)
Comment (hectcastro) [inline src/nyc_trees/apps/event/templates/event/email_rsvps.html]: It opens the model, but also throws that error in the console.

Comment (kdeloach) [inline src/nyc_trees/apps/event/templates/event/email_rsvps.html]: I see, that's very strange.

File: src/nyc_trees/apps/core/models.py (hunk 2 above)
Comment (kdeloach) [inline src/nyc_trees/apps/core/models.py]: Removed.

File: src/nyc_trees/apps/event/views.py [hunk 4]
```diff
@@ -429,3 +445,26 @@ def increase_rsvp_limit(request, event_slug):
     return {
         'max_attendees': event.max_attendees
     }
+
+
+@transaction.atomic
+def event_email_unsubscribe(request, event_slug, token):
```
Comment (hectcastro) [inline src/nyc_trees/apps/event/views.py]: I think there needs to be a test in here somewhere to confirm that the username coming back with the token is the user from the current session. If I click a valid unsubscribe link of another user while logged in as another, it unsubscribes them.

File: src/nyc_trees/apps/mail/libs.py (hunk 3 above)
Comment (hectcastro) [inline src/nyc_trees/apps/mail/libs.py]: Nevermind. Apparently this works without the BCC argument.

**Comment by hectcastro** [comment]
Unsubscribe flow works other than the hitch I ran into by accidentally clicking the unsubscribe link of an RSVP as the event organizer, which erroneously unsubscribed the person who RSVP'd.

`Reply-To` header is correctly being applied to messages, and the `From` block includes the event contact name. This was also tested manually in EC2 to ensure that SES is not using its whitelist to filter outbound messages based on `Reply-To`.

File: src/nyc_trees/apps/event/views.py (hunk 4 above)
Comment (kdeloach) [inline src/nyc_trees/apps/event/views.py]: Valid concern, but this is by design. If we check against the current user session, that assumes the user must login first, which defeats the purpose of having a one-click link to unsubscribe. It's also pretty convenient during testing. Hope that makes sense.

**Comment by kdeloach** [comment]
Thanks for testing. Since event organizers have no way of obtaining user tokens in production, I think that case will not be a cause for concern.

**Comment by kdeloach** [comment]
@RickMohr Could use another pair of eyes on this one.

**Comment by kdeloach** [comment]
Reopened as #1674 to target release branch.

**Pull request closed without merging**